File: solutions/binary-tree-preorder-traversal/binary-tree-preorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */

// Goal = Return a list of values using pre-order traversal of a tree.

// store values in list
// create function for tree traversal
    // base case: if node doesn't exist, then return
    // push curr node value to list
    // traverse left child if it exists
    // traverse right child if it exists
// return list

var preorderTraversal = function(root) {
    let preorder = [];
    
    (function traversal(node = root) {
        if (!node) {
            return;
        }       
        
        preorder.push(node.val);
        
        if (node.left) {
            traversal(node.left);
        }
        
        if (node.right) {
            traversal(node.right);
        }
        
        return;
    })();
    
    return preorder;
};

// Time complexity = O(N)
// Space complexity = O(N)

File: solutions/binary-tree-preorder-traversal/binary-tree-preorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# Goal = Return a list of values using pre-order traversal of a tree.

# store values in list
# create function for tree traversal
#     base case: if node doesn't exist, then return
#     push curr node value to list
#     traverse left child if it exists
#     traverse right child if it exists
# return list

class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        preorder = []
        
        def traversal(self, node: Optional[TreeNode]) -> None:
            if node is None:
                return
            
            preorder.append(node.val)
            
            if node.left is not None:
                traversal(self, node.left)
                
            if node.right is not None:
                traversal(self, node.right)
                
            return
        
        traversal(self, root)
        
        return preorder
        
# Time complexity = O(N)
# Space complexity = O(N)
